Feature/add fedavg metric optrimization controller

rbagan opened the pull request:
Fixes # .

### Description

Hi all,

This pull request adds a new controller: **FedAvg Metric Optimizaiton**. This controller is the result of a joint effort between Roche and Universitätsspital Zurich (USZ) to train a model in a federated manner.

Purpose of the controller:
The goal is to obtain the best possible model by optimizing a specific metric (e.g., minimizing the loss or maximizing the F-Score) and to stop the training if the tracked metric does not improve after a certain number of FL rounds, as defined by the researcher. This approach saves computation time during FL training, especially when the model is large and requires a significant amount of data. This controller has been developed in a paper that is currently under peer review. 

Additionally, we wanted to provide the option to choose whether to optimize the metric during training or validation.

I would like to highlight that this contribution is thanks to Roche and the Universitätsspital Zurich (USZ)

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

holgerroth commented on examples/advanced/metric_optrimization/README.md:
```diff
@@ -0,0 +1,37 @@
+    - `task_validation_name`: specifies the name of the validation task
+    - `task_to_optimize`: indicates whether to apply metric optimization to the training or validation task
+    - `patience`: defines the number of FL rounds to wait without improvement before stopping the training
+* Model Selection: As and alternative to using a IntimeModelSelector componenet for model selection, we instead compare the metrics of the models in the workflow to select the best model each round.
```
typo: `componenet` -> `component`

holgerroth commented on nvflare/app_opt/pt/fedavg_metric_optimization.py:
```diff
@@ -0,0 +1,284 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Please use 2025 in the license files.

holgerroth commented on examples/advanced/metric_optrimization/src/cifar_fl.py:
```diff
@@ -0,0 +1,208 @@
+# (optional) set a fix place so we don't need to download everytime
+CIFAR10_ROOT = "data/cifar10"
+# (optional) We change to use GPU to speed things up.
+# if you want to use CPU, change DEVICE="cpu"
```
We typically use this so the code automatically works with CPU or GPU
```
# If available, we use GPU to speed things up.
DEVICE = "cuda" if torch.cuda.is_available() else "CPU"
```

rbagan commented:
Dear @chesterxgchen and @holgerroth,

I have updated the pull request with the reworked implementation. I believe these changes improve the Early Stopping module and add new features, including `patience` and `task_to_optimize`.

- patience (optional): Defines the number of checks with no improvement after which federated learning will stop. If set to None, this option has no effect.

- task_to_optimize (optional): Allows the user to specify where early stopping should be applied—training or validation. The task name is passed as an argument to the `send_model_and_wait` function. By default, it is set to "train."

During this rework, I also fixed a critical bug: previously, `best_model = curr_model` caused both variables to reference the same object in memory. As a result, when evaluating the stop condition, comparisons were always made against the same variable, making the condition always True. To address this, I minimized changes and avoided relying on a separate `best_model` object, thus maintaining efficiency without introducing a deep copy operation.

Breaking changes:

- Removed the class method `select_best_model`.

- Changed the signature of the `is_curr_model_better` method.

My intention was to avoid breaking changes, but introducing a deep copy for `best_model` would have added unnecessary computational overhead.

I have manually tested the code to ensure it works as expected.

I hope these changes meet your expectations. I’m happy to hear your feedback.

Thank you.

rbagan commented:
Dear @chesterxgchen ,

Thank you for your feedback. I have updated the ipynb to reflect the latest changes. Regarding the inclusion of new logic for more complex expressions, we can create a separate pull request for this. It would require additional effort and in-depth testing to ensure everything works as expected

rbagan commented:
Dear @chesterxgchen ,

Updated, thanks!
